### PR TITLE
Use go-ethereum/rpc

### PR DESCRIPTION
**Pull request by metachris**
switch from gorilla/mux to go-ethereum/rpc

### CODE REVIEW

File: lib/rpcserver.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+			Public:    true,
+		},
+	}
+	if err := node.RegisterApis(apis, []string{namespace}, srv, false); err != nil {
+		return nil, fmt.Errorf("could not register api: %s", err)
+	}
+	return srv, nil
+}
+
+// NewHTTPServer TODO
```
Comment (lightclient) [inline lib/rpcserver.go:79]: Is this TODO for a comment or that this function needs work?

File: lib/service.go [hunk 2]
```diff
@@ -29,17 +28,23 @@ type BoostService struct {
...
 }
 
-func newBoostService(relayURLs []string, log *logrus.Entry) (*BoostService, error) {
+func newBoostService(relayURLs []string, log *logrus.Entry, getHeaderTimeout time.Duration) (*BoostService, error) {
 	if len(relayURLs) == 0 || relayURLs[0] == "" {
 		return nil, errors.New("no relayURLs")
 	}
 
+	// Set custom GetHeader timeout
+	_getHeaderTimeout := defaultGetHeaderTimeout
```
Comment (lightclient) [inline lib/service.go:37]: Seems like you could get rid of this assignment by directly setting `getHeaderTimeout` if it is `0`?
```go
if getHeaderTimeout == 0 {
    getHeaderTimeout = defaultGetHeaderTimeout
}
```

**Review by lightclient — CHANGES_REQUESTED** [review]
LGTM

File: lib/service.go [hunk 3]
```diff
@@ -29,17 +28,23 @@ type BoostService struct {
...
+	if getHeaderTimeout > 0 {
+		_getHeaderTimeout = getHeaderTimeout
+	}
+
 	return &BoostService{
 		relayURLs: relayURLs,
 		log:       log.WithField("prefix", "lib/service"),
 
 		httpClient:      http.Client{Timeout: defaultHTTPTimeout},
```
[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:46]: I don't think it's required. Why would it be?

File: lib/service.go [hunk 4]
```diff
@@ -118,7 +119,7 @@ func (m *BoostService) SetFeeRecipientV1(_ *http.Request, args *[]interface{}, r
 			}
 
 			// Result should be true if any one relay responds true
-			*result = *result || _result
+			result = result || _result
```
[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:122]: Nope, they don't share a variable, the goroutines handle independent RPC requests and independent data.

File: lib/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhakiyakalimuthu) [reply lib/service.go:46]: As per go net/http client, has to be a pointer. is there any reason ? :thinking: https://github.com/golang/go/blob/master/src/net/http/client.go#L586

File: lib/service.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhakiyakalimuthu) [reply lib/service.go:122]: https://github.com/flashbots/mev-boost/pull/93/files#diff-c48208726be97e20698f1bd874d4bcece8e2af190e67fe6c1578aae3ba212bf4R93
Basically 'result` is a shared variable right ?

[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:122]: No, it's not shared. One instance per request.

File: lib/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:46]: The line you like to does not mean that a client instance has to be referenced with a pointer. How your reference it does not matter at all for the instance itself.

[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:46]: The code you linked shows that the client methods are declared with pointer receivers, which means they can modify the underlying instance. https://go.dev/tour/methods/4

[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go]: Accidentally applied this change due to fat fingers on the phone :-(

File: lib/service.go [hunk 5]
```diff
@@ -81,22 +86,18 @@ type rpcResponseContainer struct {
 }
 
 // SetFeeRecipientV1 - returns true if at least one relay returns true
-func (m *BoostService) SetFeeRecipientV1(_ *http.Request, args *[]interface{}, result *bool) error {
+func (m *BoostService) SetFeeRecipientV1(ctx context.Context, feeRecipient, timestamp, publicKey, signature string) (*bool, error) {
```
[the remark this answers is not in the document]
Comment (metachris) [reply lib/service.go:89]: used now, good catch 👍

File: lib/rpcserver.go (hunk 1 above)
Comment (metachris) [reply lib/rpcserver.go]: fixed, was just the comment to make the linter happy

File: lib/service.go (hunk 2 above)
Comment (metachris) [reply lib/service.go]: nice, done 🙏

File: lib/rpcserver.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	apis := []gethRpc.API{
+		{
+			Namespace: namespace,
+			Version:   "1.0",
+			Service:   backend,
+			Public:    true,
+		},
+	}
+	if err := node.RegisterApis(apis, []string{namespace}, srv, false); err != nil {
+		return nil, fmt.Errorf("could not register api: %s", err)
```
Comment (terencechain) [inline lib/rpcserver.go:74]: Have we considered `%w` verb which supports wrapping?

```suggestion
		return nil, fmt.Errorf("could not register api: %w", err)
```

Comment (metachris) [reply lib/rpcserver.go]: Right, makes sense! Will keep that in mind :)

File: lib/service.go [hunk 7]
```diff
@@ -127,32 +127,33 @@ func (m *BoostService) SetFeeRecipientV1(_ *http.Request, args *[]interface{}, r
...
 	method := "builder_getHeaderV1"
 	logMethod := m.log.WithField("method", method)
 
 	if len(*blockHash) != 66 {
-		return fmt.Errorf("invalid block hash: %s", *blockHash)
+		return nil, fmt.Errorf("invalid block hash: %s", *blockHash)
 	}
 
 	// Call the relay
+	result := new(GetHeaderResponse)
```
Comment (jparyani) [inline lib/service.go:149]:
```suggestion
	var result GetHeaderResponse
```
nitpick: unnecessary allocation, could be replaced by a `var` declaration

**Review by jparyani — APPROVED** [review]

File: lib/service.go (hunk 7 above)
Comment (metachris) [reply lib/service.go:149]: What would be the difference? `var` also allocates the memory 🤔

Comment (jparyani) [reply lib/service.go:149]: This is a pointer variable, you’re allocating the value struct unnecessarily

Comment (jparyani) [reply lib/service.go:149]: Ah my bad, the code suggestion missed the pointer, my bad :)

**Pull request merged**